### PR TITLE
consensus: check block parts don't exceed maximum block bytes

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Closes: #5423

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -1779,16 +1779,23 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
 	if err != nil {
 		return added, err
 	}
+	if cs.ProposalBlockParts.ByteSize() > cs.state.ConsensusParams.Block.MaxBytes {
+		return added, fmt.Errorf("total size of proposal block parts exceeds maximum block bytes(%d > %d)",
```
Comment (melekes) [inline consensus/state.go:1783]:
```suggestion
		return added, fmt.Errorf("total size of proposal block parts exceeds maximum block bytes (%d > %d)",
```

File: types/part_set.go [hunk 2]
```diff
@@ -155,6 +155,9 @@ type PartSet struct {
 	parts         []*Part
 	partsBitArray *bits.BitArray
 	count         uint32
+	// a count of the total size (in bytes). Used to ensure that the
+	// part set doesn't exceed the maximum block bytes
+	byteSize int64
```
Comment (melekes) [inline types/part_set.go:160]: why int64?

Comment (cmwaters) [reply types/part_set.go:160]: it matches MaxBytes in the Block Params which is also int64. tbh it's still a relatively arbitrary number

File: consensus/state.go [hunk 3]
```diff
@@ -1779,16 +1779,23 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
 	if err != nil {
 		return added, err
 	}
+	if cs.ProposalBlockParts.ByteSize() > cs.state.ConsensusParams.Block.MaxBytes {
```
Comment (marbar3778) [inline consensus/state.go:1782]: Could the same thing be achieved by doing?
```
reader := io.LimitReader(
cs.ProposalBlockParts.GetReader(),
cs.state.ConsensusParams.Block.MaxBytes)
bz, err := ioutil.ReadAll(reader)
```

Comment (cmwaters) [reply consensus/state.go:1782]: Is this more performant?

Comment (marbar3778) [reply consensus/state.go:1782]: I am not sure if its more performant but causes fewer changes and I believe mimics the previous behavior. I would have to check what amino was doing.

**Review by melekes — APPROVED** [review]

**Comment by melekes** [comment]
could u also cheer-pick https://github.com/tendermint/tendermint/commit/433bdf5063052ec2f0b93e895fc937c94088b02b onto v0.34.0-rc5 branch?

File: types/part_set.go (hunk 2 above)
Comment (erikgrinaker) [inline types/part_set.go:160]: It would've been even simpler to just calculate the size on the fly in `ByteSize()`, this isn't performance critical.

Comment (cmwaters) [reply types/part_set.go:160]: Yeah true but it's a relatively minor difference right?

Comment (erikgrinaker) [reply types/part_set.go:160]: Just less stuff that can go wrong :)

File: types/part_set.go [hunk 4]
```diff
@@ -186,6 +189,7 @@ func NewPartSetFromData(data []byte, partSize uint32) *PartSet {
 		parts:         parts,
 		partsBitArray: partsBitArray,
 		count:         total,
+		byteSize:      int64(len(data)),
```
Comment (marbar3778) [inline types/part_set.go:192]: i think this is culprit, you should be accounting for the parts not the entire struct

**Pull request merged**